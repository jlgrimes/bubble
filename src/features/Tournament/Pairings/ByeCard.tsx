import Typography from "@mui/material/Typography";
import { PairingHeaderCard } from "./PairingHeaderCard";

export const ByeCard = () => {
  return (
    <PairingHeaderCard aria-label='Bye'>
      <Typography>Bye</Typography>
    </PairingHeaderCard>
  );
};