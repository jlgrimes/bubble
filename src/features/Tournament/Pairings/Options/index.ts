import { PairingFilters } from './PairingFilters';
import { PairingSearch } from './PairingSearch';

export { PairingFilters, PairingSearch };
