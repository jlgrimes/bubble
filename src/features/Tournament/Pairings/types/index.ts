import { Match } from "./Match";
import { MatchResult } from "./MatchResult";

export type { Match, MatchResult };