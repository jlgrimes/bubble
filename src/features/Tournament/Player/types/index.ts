import { Player } from "./Player";
import { PlayerMatch } from "./PlayerMatch";
import { Record } from "./Record";

export type { Player, PlayerMatch, Record };