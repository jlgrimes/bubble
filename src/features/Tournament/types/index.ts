import { TournamentState } from "./Tournament";

export type { TournamentState };