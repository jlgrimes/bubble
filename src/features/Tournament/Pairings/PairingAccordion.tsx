import React from 'react';
import Accordion from '@mui/material/Accordion';
import MuiAccordionSummary from '@mui/material/AccordionSummary';
import MuiAccordionDetails from '@mui/material/AccordionDetails';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import styled from '@emotion/styled';

import { PairingHeaderProps, PairingHeader } from './PairingHeader';
import { PairingButtons } from './PairingButtons';
import { Match } from './types';

interface PairingAccordionProps extends PairingHeaderProps {
  /**
   * The completed match if the match was already reported.
   */
  completedMatch?: Match;
  /**
   * If the accordion is expanded. Controlled by if any other accordion is expanded.
   */
  expanded: boolean;
  /**
   * handle change when the accordion
   */
  handleChange: Function;
}

const AccordionSummary = styled(MuiAccordionSummary)`
  min-height: 0;
  padding: 0;
  &.Mui-expanded {
    min-height: 0;
  }

  .MuiAccordionSummary-content,
  .MuiAccordionSummary-content.Mui-expanded {
    margin: 0;
  }
`;

const AccordionDetails = styled(MuiAccordionDetails)`
  padding: 0;
`;

const PairingAccordion = (props: PairingAccordionProps) => {
  const { expanded, handleChange, ...rest } = props;
  const [renderAccordionDetails, setRenderAccordionDetails] = React.useState<boolean>(false);
  const [accordionExpanded, setAccordionExpanded] = React.useState<boolean>(false);

  React.useEffect(() => {
    if (expanded === false) {
      setAccordionExpanded(false);
      const timer = setTimeout(() => {
        setRenderAccordionDetails(false);
      }, 300);
      return () => clearTimeout(timer)
    } else {
      setRenderAccordionDetails(true);
      setAccordionExpanded(true);
    }
  }, [expanded]);

  const pairingIsBye: boolean = props.secondPlayer.id === 'bye';

  return (
    <Accordion
      expanded={accordionExpanded}
      onChange={!pairingIsBye && handleChange()}
      classes={{ root: 'pairing-accordion' }}
    >
      <AccordionSummary aria-controls='panel1a-content' id='panel1a-header'>
        <PairingHeader {...rest} />
      </AccordionSummary>
      {renderAccordionDetails && !pairingIsBye && (
        <AccordionDetails>
          <PairingButtons
            firstPlayer={props.firstPlayer}
            secondPlayer={props.secondPlayer}
            completedMatch={props.completedMatch}
          />
        </AccordionDetails>
      )}
    </Accordion>
  );
};

export default PairingAccordion;
