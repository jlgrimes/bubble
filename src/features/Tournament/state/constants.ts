import { Player } from '../Player/types';

export const byePlayer: Player = {
  id: 'bye',
  name: 'Bye',
  matches: [],
  record: {
    wins: 0,
    ties: 0,
    losses: 0,
  },
  matchPoints: 0,
};
