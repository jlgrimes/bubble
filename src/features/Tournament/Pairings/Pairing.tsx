import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import PairingAccordion from './PairingAccordion';
import { RootState } from '../../../app/store';
import type { Match } from './types';
import type { Player } from '../Player/types';
import { alterWithCompletedMatch } from '../Player/utils/player';
import { submitMatchResult } from '../state/tournamentSlice';
import { getExistingMatch } from './utils/match';

interface PairingProps {
  pairing: string[];
  idx: number;
  expandedPairing: number | boolean;
  setExpandedPairing: Function;
}

export const Pairing = (props: PairingProps) => {
  const dispatch = useDispatch();

  const round: number = useSelector(
    (state: RootState) => state.tournament.round
  );
  const players: Player[] = useSelector(
    (state: RootState) => state.tournament.players
  );
  const matchResults: Match[] = useSelector(
    (state: RootState) => state.tournament.matchResults
  );

  const existingMatch = getExistingMatch(props.pairing, matchResults);
  // TODO: error handing for find?
  const firstPlayer: Player = players.find(
    player => player.id === props.pairing[0]
  )!;
  const secondPlayer: Player = players.find(
    player => player.id === props.pairing[1]
  )!;

  return (
    <PairingAccordion
      completedMatch={existingMatch}
      expanded={props.expandedPairing === props.idx}
      handleChange={() => (event: React.SyntheticEvent, isExpanded: boolean) => {
        props.setExpandedPairing(isExpanded ? props.idx : false);
      }}
      firstPlayer={alterWithCompletedMatch(firstPlayer, existingMatch)}
      secondPlayer={alterWithCompletedMatch(secondPlayer, existingMatch)}
      table={props.idx + 1}
    />
  );
};