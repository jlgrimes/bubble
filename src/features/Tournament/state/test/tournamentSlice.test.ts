import { generateEmptyPlayers, SAMPLE_PLAYER } from '../../../../helpers/testConstants';
import { Match } from '../../Pairings/types';
import { byePlayer } from '../constants';
import reducer, {
  addPlayer,
  nextRound,
  removePlayer,
} from '../tournamentSlice';
import { TournamentState } from '../TournamentState';
import { ViewState } from '../ViewState';

describe('tournament reducers', () => {
  const initialState: TournamentState = {
    round: 0,
    pairings: [],
    players: [],
    matchResults: [],
    maxRounds: 5,
    topCut: undefined,
    viewState: 'tournament'
  }

  describe('addPlayer', () => {
    it('should add player', () => {
      expect(reducer(initialState, addPlayer(SAMPLE_PLAYER))).toEqual({
        ...initialState,
        players: [SAMPLE_PLAYER],
      });
    });
  });

  describe('removePlayer', () => {
    it('should remove player', () => {
      const previousState = {
        ...initialState,
        players: [SAMPLE_PLAYER],
      };
      expect(reducer(previousState, removePlayer('0'))).toEqual(initialState);
    });
  });

  // describe('nextRound', () => {
  //   it('should generate next round pairings with bye', () => {
  //     const players = [...generateEmptyPlayers(5), byePlayer];
  //     const previousState = {
  //       round: 1,
  //       pairings: [['0', '1'], ['2', '3'], ['4', 'bye']],
  //       players,
  //       matchResults: [
  //         { playerIds: ['0', '1'], result: 'win' } as Match,
  //         { playerIds: ['2', '3'], result: 'win' } as Match,
  //         { playerIds: ['4', 'bye'], result: 'win' } as Match,
  //       ],
  //       deterministicPairing: true,
  //       maxRounds: 5,
  //       topCut: undefined,
  //       viewState: 'tournament' as ViewState
  //     };
  //     expect(reducer(previousState, nextRound())).toMatchSnapshot();
  //   });
  // });
});
