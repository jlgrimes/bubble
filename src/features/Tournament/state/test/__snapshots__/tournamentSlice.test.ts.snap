// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tournament reducers nextRound should generate next round pairings with bye 1`] = `
Object {
  "matchResults": Array [],
  "pairings": Array [
    Array [
      "4",
      "3",
    ],
    Array [
      "0",
      "2",
    ],
    Array [
      "1",
    ],
  ],
  "players": Array [
    Object {
      "id": "0",
      "matchPoints": 3,
      "matches": Array [
        Object {
          "opponentId": "1",
          "result": "win",
        },
      ],
      "name": "Player 0",
      "record": Object {
        "losses": 0,
        "ties": 0,
        "wins": 1,
      },
    },
    Object {
      "id": "1",
      "matchPoints": 0,
      "matches": Array [
        Object {
          "opponentId": "0",
          "result": "loss",
        },
      ],
      "name": "Player 1",
      "record": Object {
        "losses": 1,
        "ties": 0,
        "wins": 0,
      },
    },
    Object {
      "id": "2",
      "matchPoints": 3,
      "matches": Array [
        Object {
          "opponentId": "3",
          "result": "win",
        },
      ],
      "name": "Player 2",
      "record": Object {
        "losses": 0,
        "ties": 0,
        "wins": 1,
      },
    },
    Object {
      "id": "3",
      "matchPoints": 0,
      "matches": Array [
        Object {
          "opponentId": "2",
          "result": "loss",
        },
      ],
      "name": "Player 3",
      "record": Object {
        "losses": 1,
        "ties": 0,
        "wins": 0,
      },
    },
    Object {
      "id": "4",
      "matchPoints": 3,
      "matches": Array [
        Object {
          "opponentId": undefined,
          "result": "win",
        },
      ],
      "name": "Player 4",
      "record": Object {
        "losses": 0,
        "ties": 0,
        "wins": 1,
      },
    },
  ],
  "round": 2,
}
`;
